### ISSUE TITLE
Error if target consists of only background class

### ISSUE THREAD

**Issue by Harut0726** [XuJiacong/PIDNet#14]
Hi, thank you for sharing your work. I tried to train the model on my custom dataset derived from CamVid and I'm getting error during calculating the loss function when target consists of only background class (255). The error appears on 73th line in utils/criterion.py.
```
pred = pred.gather(1, tmp_target.unsqueeze(1))
pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
min_value = pred[min(self.min_kept, pred.numel() - 1)] # --- here is the error because pred is an empty array and mask only consist of False values
threshold = max(min_value, self.thresh)
 ```

**Comment by Daniel595** [comment]
I got the same error. Any updates on this?

**Comment by tomihisaw** [comment]
I also noticed this error. My current workaround:
```
       if pred.numel() > 0:
            min_value = pred[min(self.min_kept, pred.numel() - 1)]
            threshold = max(min_value, self.thresh)
        else:
            threshold = self.thresh
```

**Comment by m828** [comment]
I also encountered this problem, is there a solution? (The number of backgrounds is still a lot of datasets)

**Comment by Daniel595** [comment]
Did you try @tomihisaw's workaround? It works fine for me. Nearly 90% of my custom-dataset is background only and my training results are sufficient. The output during training didn't look very good to me (might be caused generally by inbalance) but the final prediction and visualization on a seperate test are ok.

**Comment by m828** [comment]
> Did you try @tomihisaw's workaround? It works fine for me. Nearly 90% of my custom-dataset is background only and my training results are sufficient. The output during training didn't look very good to me (might be caused generally by inbalance) but the final prediction and visualization on a seperate test are ok.

Oh? I can run it, but it still doesn't work after a few epochs. I will not continue, then I will try again, thank you.
![image](https://user-images.githubusercontent.com/45037914/183840302-7f0b23c1-55f2-4a1e-87e7-8a32cc813c84.png)

**Comment by m828** [comment]
> Did you try @tomihisaw's workaround? It works fine for me. Nearly 90% of my custom-dataset is background only and my training results are sufficient. The output during training didn't look very good to me (might be caused generally by inbalance) but the final prediction and visualization on a seperate test are ok.

I tried to run the model, but the mIoU is still 0.1250, how is your data file written?

**Comment by Daniel595** [comment]
I'm using my own dataset class with pandas dataframe. So I think my data file would not help you here. 
On your output it looks like all other classes were missing. Did you doublecheck the segmentation masks?
Loss nan happened to me if the learning rate was to high (like x10, x100).
I also disabled multiscale since I didn't find any parameters that worked well.
Maybe there is sth. helpfull.

**Comment by scl666** [comment]
> 我也注意到这个错误。我目前的解决方法：
> 
> ```
>        if pred.numel() > 0:
>             min_value = pred[min(self.min_kept, pred.numel() - 1)]
>             threshold = max(min_value, self.thresh)
>         else:
>             threshold = self.thresh
> ```
> 
> 对我来说，这是一个不常见的错误，因此解决方法似乎不会对训练产生负面影响。但是，如果您有许多带有无效/忽略像素的图像，则可能不是正确的解决方案。

你好，我改完之后损失变成了nan，大佬方便发下具体文件，QQ：2358919383，邮箱：2358919283@qq.com

**Comment by scl666** [comment]
> Jul

你好，请问这个错误找到解决方法了吗？我也遇到这个问题了，想请教一下

**Comment by QiaoShin** [comment]
> Hi, thank you for sharing your work. I tried to train the model on my custom dataset derived from CamVid and I'm getting error during calculating the loss function when target consists of only background class (255). The error appears on 73th line in utils/criterion.py.
> 
> ```
> pred = pred.gather(1, tmp_target.unsqueeze(1))
> pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
> min_value = pred[min(self.min_kept, pred.numel() - 1)] # --- here is the error because pred is an empty array and mask only consist of False values
> threshold = max(min_value, self.thresh)
> ```



> > Jul
> 
> 你好，请问这个错误找到解决方法了吗？我也遇到这个问题了，想请教一下

这个问题在难例挖掘ohem loss里几乎都会遇到。
此处你们出现此问题主要出现在这一行代码中：https://github.com/XuJiacong/PIDNet/blob/f0ac91cdea7bf0cb2077b65e960c5b98b9173b0f/utils/utils.py#L53
code：`bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.8, labels, filler)`
注意，此处的0.8位作者设置的难例阈值，ohem官方给的阈值为0.7。当bd_lael中元素值全为1，即表示outputs的第三张张量（分割头d的输出）在0.8阈值情况下没有难例，这种情况下传输到ohem的forward中计算，就会出现pred为空的情况。

我的理解是，如果无难例情况，不如直接将整个分割头d的输出输入ohem loss中进行计算，改动如下：
`    try:
        bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.7, labels, filler)
        loss_sb = self.sem_loss([outputs[-2]], bd_label)
    except:
        loss_sb = self.sem_loss([outputs[-2]], labels)`
（此处阈值我使用的官方0.7）

**Comment by 17648145240** [comment]
> 

Hello, buddy. I have encountered the same problem as you. Have you finally solved this problem? If so, can you share your modified code with me? My email address is: 202131903057@stu.hebut.edu.cn

**Comment by 17648145240** [comment]
> > Did you try @tomihisaw's workaround? It works fine for me. Nearly 90% of my custom-dataset is background only and my training results are sufficient. The output during training didn't look very good to me (might be caused generally by inbalance) but the final prediction and visualization on a seperate test are ok.
> 
> Oh? I can run it, but it still doesn't work after a few epochs. I will not continue, then I will try again, thank you. ![image](https://user-images.githubusercontent.com/45037914/183840302-7f0b23c1-55f2-4a1e-87e7-8a32cc813c84.png)

Hello, buddy. I have encountered the same problem as you. Have you finally solved this problem? If so, can you share your modified code with me? My email address is: 202131903057@stu.hebut.edu.cn

**Comment by Unusual6** [comment]
> > > Did you try @tomihisaw's workaround? It works fine for me. Nearly 90% of my custom-dataset is background only and my training results are sufficient. The output during training didn't look very good to me (might be caused generally by inbalance) but the final prediction and visualization on a seperate test are ok.
> > 
> > 
> > Oh? I can run it, but it still doesn't work after a few epochs. I will not continue, then I will try again, thank you. ![image](https://user-images.githubusercontent.com/45037914/183840302-7f0b23c1-55f2-4a1e-87e7-8a32cc813c84.png)
> 
> Hello, buddy. I have encountered the same problem as you. Have you finally solved this problem? If so, can you share your modified code with me? My email address is: [202131903057@stu.hebut.edu.cn](mailto:202131903057@stu.hebut.edu.cn)

I have the same error. But all the solution above cannot solve my problem. I find the reason that internet have a lot of version of camvid（lable is colored or black-white）,if you use colored version that can solve this problem and train normally.

make sure your camvid label like this:

![image](https://github.com/XuJiacong/PIDNet/assets/114602129/9ec67b08-cc98-438e-95b5-8f2139219046)
 
instead of

![image](https://github.com/XuJiacong/PIDNet/assets/114602129/7e3d285d-a99a-4b8f-93f0-fe4a772b746a)

**Comment by Unusual6** [comment]
I have the same error. But all the solution above cannot solve my problem. I find the reason that internet have a lot of version of camvid（lable is colored or black-white）,if you use colored version that can solve this problem and train normally.

make sure your camvid label like this:

![image](https://github.com/XuJiacong/PIDNet/assets/114602129/9ec67b08-cc98-438e-95b5-8f2139219046)
 
instead of

![image](https://github.com/XuJiacong/PIDNet/assets/114602129/7e3d285d-a99a-4b8f-93f0-fe4a772b746a)

**Comment by stemper0123** [comment]
> I have the same error. But all the solution above cannot solve my problem. I find the reason that internet have a lot of version of camvid（lable is colored or black-white）,if you use colored version that can solve this problem and train normally.

Did you modify the __getitem__ function within the CamVid dataset class? The original CamVid dataset class uses the color2label method, which translates the colored map into grayscale label IDs. This may have potentially resulted with the images that are already in grayscale in weird class IDs that are outside the range. I have mentioned it in https://github.com/XuJiacong/PIDNet/issues/65.

**Comment by wakakkakak** [comment]
> > Hi, thank you for sharing your work. I tried to train the model on my custom dataset derived from CamVid and I'm getting error during calculating the loss function when target consists of only background class (255). The error appears on 73th line in utils/criterion.py.
> > ```
> > pred = pred.gather(1, tmp_target.unsqueeze(1))
> > pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
> > min_value = pred[min(self.min_kept, pred.numel() - 1)] # --- here is the error because pred is an empty array and mask only consist of False values
> > threshold = max(min_value, self.thresh)
> > ```
> 
> > > Jul
> > 
> > 
> > 你好，请问这个错误找到解决方法了吗？我也遇到这个问题了，想请教一下
> 
> 这个问题在难例挖掘ohem loss里几乎都会遇到。 此处你们出现此问题主要出现在这一行代码中：
> 
> https://github.com/XuJiacong/PIDNet/blob/f0ac91cdea7bf0cb2077b65e960c5b98b9173b0f/utils/utils.py#L53
> 
> 
> code：`bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.8, labels, filler)`
> 注意，此处的0.8为作者设置的难例阈值，ohem官方给的阈值为0.7。当bd_label中元素值全为1，即表示outputs的第三张张量（分割头d的输出）在0.8阈值情况下没有难例，这种情况下传输到ohem的forward中计算，就会出现pred为空的情况。
> 我的理解是，如果无难例情况，替换成将整个labels输入ohem loss中进行计算，改动如下：
> 
> ```
> try:
>     bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.7, labels, filler)
>     loss_sb = self.sem_loss([outputs[-2]], bd_label)
> except:
>     loss_sb = self.sem_loss([outputs[-2]], labels)
> ```
> 
> （此处阈值我使用的官方0.7）

![PID1](https://github.com/XuJiacong/PIDNet/assets/117575513/114dc6a8-40b9-400b-acbb-42bcb3740a02)
我修改后（将0.8改为0.7）还是报一样的错，请问是原因呢？

**Comment by fatejzz** [comment]
> > > Hi, thank you for sharing your work. I tried to train the model on my custom dataset derived from CamVid and I'm getting error during calculating the loss function when target consists of only background class (255). The error appears on 73th line in utils/criterion.py.
> > > ```
> > > pred = pred.gather(1, tmp_target.unsqueeze(1))
> > > pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
> > > min_value = pred[min(self.min_kept, pred.numel() - 1)] # --- here is the error because pred is an empty array and mask only consist of False values
> > > threshold = max(min_value, self.thresh)
> > > ```
> > 
> > 
> > > > Jul
> > > 
> > > 
> > > 你好，请问这个错误找到解决方法了吗？我也遇到这个问题了，想请教一下
> > 
> > 
> > 这个问题在难例挖掘ohem loss里几乎都会遇到。 此处你们出现此问题主要出现在这一行代码中：
> > https://github.com/XuJiacong/PIDNet/blob/f0ac91cdea7bf0cb2077b65e960c5b98b9173b0f/utils/utils.py#L53
> > 
> > code：`bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.8, labels, filler)`
> > 注意，此处的0.8为作者设置的难例阈值，ohem官方给的阈值为0.7。当bd_label中元素值全为1，即表示outputs的第三张张量（分割头d的输出）在0.8阈值情况下没有难例，这种情况下传输到ohem的forward中计算，就会出现pred为空的情况。
> > 我的理解是，如果无难例情况，替换成将整个labels输入ohem loss中进行计算，改动如下：
> > ```
> > try:
> >     bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.7, labels, filler)
> >     loss_sb = self.sem_loss([outputs[-2]], bd_label)
> > except:
> >     loss_sb = self.sem_loss([outputs[-2]], labels)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > （此处阈值我使用的官方0.7）
> 
> ![PID1](https://private-user-images.githubusercontent.com/117575513/311495434-114dc6a8-40b9-400b-acbb-42bcb3740a02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4NzQ1MjksIm5iZiI6MTcyMTg3NDIyOSwicGF0aCI6Ii8xMTc1NzU1MTMvMzExNDk1NDM0LTExNGRjNmE4LTQwYjktNDAwYi1hY2JiLTQyYmNiMzc0MGEwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyNVQwMjIzNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YmI4ZDc5ZWY5NTIxZDI5M2RiNWIxMzQzZTMyYTc4ZDZiNWYxNWMxYWVhYTJlZTM0YzNmZGY1M2I5MDQ4YWI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GD0PukRN5o5_sIFV82PVn4Qf9_aHR3XTRqyhAhaaxuY) 我修改后（将0.8改为0.7）还是报一样的错，请问是原因呢？

你好，我这边也出现了这样的问题，经过调试后发现报错原因大概是此时边缘分割输出头所有的输出均小于阈值设置，致使bd_label均为0，后续出错。根据作者原文的意思，这里应该是对应原文中的L3，那么此时我觉得可能是loss_sb为0吧，即不考虑这一部分损失，如果再用loss_sb = self.sem_loss([outputs[-2]], labels)这个可能跟之前的主分割头输出重复了

**Comment by zhoukaiyu123** [comment]
> > > 您好，感谢您分享您的工作。我尝试在从 CamVid 派生的自定义数据集上训练模型，当 target 仅包含背景类 （255） 时，我在计算损失函数时出现错误。该错误出现在 utils/criterion.py 的第 73 行。
> > > ```
> > > pred = pred.gather(1, tmp_target.unsqueeze(1))
> > > pred, ind = pred.contiguous().view(-1,)[mask].contiguous().sort()
> > > min_value = pred[min(self.min_kept, pred.numel() - 1)] # --- here is the error because pred is an empty array and mask only consist of False values
> > > threshold = max(min_value, self.thresh)
> > > ```
> > 
> > 
> > > > 七月
> > > 
> > > 
> > > 你好，请问这个错误找到解决方法了吗？我也遇到这个问题了，想请教一下
> > 
> > 
> > 这个问题在难例挖掘ohem loss里几乎都会遇到。此处你们出现此问题主要出现在这一行代码中：
> > https://github.com/XuJiacong/PIDNet/blob/f0ac91cdea7bf0cb2077b65e960c5b98b9173b0f/utils/utils.py#L53
> > 
> > code：
> > 注意，此处的0.8为作者设置的难例阈值，ohem官方给的阈值为0.7。当bd_label中元素值全为1，即表示outputs的第三张张量（分割头d的输出）在0.8阈值情况下没有难例，这种情况下传输到ohem的forward中计算，就会出现pred为空的情况。
> > 我的理解是，如果无难例情况，替换成将整个labels输入ohem loss中进行计算，改动如下：`bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.8, labels, filler)`
> > ```
> > try:
> >     bd_label = torch.where(torch.sigmoid(outputs[-1][:,0,:,:])>0.7, labels, filler)
> >     loss_sb = self.sem_loss([outputs[-2]], bd_label)
> > except:
> >     loss_sb = self.sem_loss([outputs[-2]], labels)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > （此处阈值我使用的官方0.7）
> 
> ![PID1](https://private-user-images.githubusercontent.com/117575513/311495434-114dc6a8-40b9-400b-acbb-42bcb3740a02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMzNzA5NzQsIm5iZiI6MTczMzM3MDY3NCwicGF0aCI6Ii8xMTc1NzU1MTMvMzExNDk1NDM0LTExNGRjNmE4LTQwYjktNDAwYi1hY2JiLTQyYmNiMzc0MGEwMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwNVQwMzUxMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYjAzNDI1NDU5MzdlMTg1NjI5MjEyODk0YjA4MTQwNGJjZGVkYzk0NmE0MGJiYmZmYmYwZmY5Yzc4ZmNhMzRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.t72eooOBVSYHvYR3rZw53Ztc91Cr1C5TQcrWXCzvbzk)我修改后（将0.8改为0.7）还是报一样的错，请问是原因呢？

请问您解决了吗
